feat: speed: Measure vehicle speed using pin-change IRQ on VSS

Hurricos opened the pull request:
feat: speed: Measure vehicle speed using pin-change IRQ on VSS

This commit implements vehicle speed measurement using a pin-change
interrupt on the vehicle speed sensor (VSS) line connected to pin D7

The VSS emits ~4250 pulses per mile[^1] under normal conditions. As
with NEP, attach a pin-change interrupt; if when the interrupt
activates, the pin is high, recalculate the vehicle speed by checking
the time since the last revolution.

Also, reflect the new measurement details in
`debugUSB_printOEMsignals` (the OEM signal echoing mode).

Notes:

- A more precise value can be derived from the Honda DTC manual [^2]:

  > The vehicle speed sensor (VSS) is attached to the transmission
  > housing. The VSS outputs a pulsing signal for every revolu- tion
  > of the differential gear via the speedometer gear.

[^1]:
https://www.insightcentral.net/threads/vss-vehicle-speed-sensor-output.13597/post-191883
describes it initially as 4550 pulses, but some experimentation and
comparison to a stock speedo led to the 4250 number now used.

[^2]: https://downloads.laboratoryb.org/insight/other_manuals/HONDA_DTC.pdf#page=106

Signed-off-by: Martin Kennedy <hurricos@gmail.com>

### Addenda

Note that this PR sits atop the fix-rpm PR; we'll rebase when that (as a bug-level issue) gets merged.

Hurricos committed:
feat: speed: Measure vehicle speed using pin-change IRQ on VSS

This commit implements vehicle speed measurement using a pin-change
interrupt on the vehicle speed sensor (VSS) line connected to pin D7

The VSS emits ~4250 pulses per mile[^1] under normal conditions. As
with NEP, attach a pin-change interrupt; if when the interrupt
activates, the pin is high, recalculate the vehicle speed by checking
the time since the last revolution.

Also, reflect the new measurement details in
`debugUSB_printOEMsignals` (the OEM signal echoing mode).

Notes:

- A more precise value can be derived from the Honda DTC manual [^2]:

  > The vehicle speed sensor (VSS) is attached to the transmission
  > housing. The VSS outputs a pulsing signal for every revolu- tion
  > of the differential gear via the speedometer gear.

[^1]:
https://www.insightcentral.net/threads/vss-vehicle-speed-sensor-output.13597/post-191883
describes it initially as 4550 pulses, but some experimentation and
comparison to a stock speedo led to the 4250 number now used.

[^2]: https://downloads.laboratoryb.org/insight/other_manuals/HONDA_DTC.pdf#page=106

Signed-off-by: Martin Kennedy <hurricos@gmail.com>

doppelhub commented:
A note in data I collected previously says "VSS is ~4000 pulses/mile (89 Hz@80mph)".  Does that agree with your observations?

Hurricos commented:
> A note in data I collected previously says "VSS is ~4000 pulses/mile (89 Hz@80mph)". Does that agree with your observations?

I played around, and the number that got closest was 4250.

Hurricos commented:
The force-push from 42c3cb7 to 61b4032 adjusts the VSS divisor to 4250 per our conversation.

doppelhub reviewed (changes requested):
Please define magic number (4250) and then reduce instructions in ISR.  Existing:
latestVehicleMPH = ONE_HOUR_IN_MICROSECONDS / periodBetweenTicks_us / 4250;

proposed:
latestVehicleMPH = (ONE_HOUR_IN_MICROSECONDS / PULSES_PER_MILE ) / periodBetweenTicks_us;

Hurricos commented:
Commit 8d8cacb831535a1b62334b0d903e94a04742f212 is a force-push rebasing `vehicle-speed` onto current `main`.

Hurricos committed:
cleanup/optimize: #define magic number; group constants

Should have dealt with defining magic number in advance. Oops. It's
fine, though, we can poke this number when someone realizes it's off
by number%.

Again, as per JTS, the preprocessor will perform all constant-only
math if it's all grouped together.

Grouping the constants together reduces the length fo the ISR by 10
bytes[^1] ... don't be fooled, though, it skips a whole emulated
floating-point division.

Signed-off-by: Martin Kennedy <hurricos@gmail.com>

[^1]: https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1AB9U8lJL6yAngGVG6AMKpaAVxYM9DgDJ4GmADl3ACNMYhAAdlIAB1QFQlsGZzcPPVj4mwFffyCWUPCoy0xrRKECJmICZPdPLgtMK0yGMoqCbMCQsMiLcsrq1LqFXra/DryuiIBKC1RXYmR2DgBSACYAZj9kNywAaiW1xyH8QQA6BH3sJY0AQSvb9awafx2ACQBJAHEXna47zFUCGEGDtXH4CGsVsYCDtgNFbMYAGpCITGYCYAjKPwtQEQSb7ABCfwBQJBYIhUJ2LDwyGIcVxBLuADcXEwbPRSYIuAA2CkGQFDBGYBDU%2BgAWWU332ABEdhoGTdVmtHqMdgB5ALYYwvVVyABKxjeAWMorejl1qqE2Ec6qlQh2ay5Gg02C5dyZqDw6B2THpCoihJuew0AE4hqzqRzwZDoYyFAoACrUgDWxmixEwjNsrgUxmzezWMrlawDtwVIbwVGgcNQiORqPRmOa5Rxk3zUulr0%2BLzxZeDS39bpDV2DoME5JjccTyBTDFQAHdcwo25TqbSFPTi4PS32Q6OoxTomFbOh8ei55hGFOkzm8x3Ywnk8ZZwu8wBaHb3q%2Bp9OZ2Y3hTyqW27DnymACkKIqYOKkoFmqGpajq%2BqGsaprmpa1oBLaOwAPQ7IexDHqeBDnpeyb/jhOySCsACsRYloOO7Bp%2Bj5phmWbkXek6Ps%2Bi6AcO/btn6UocNMtCcNRvCeNwvCoJwjg7AoszzJgezrDwpAEJoInTEmIA0fonCSJJWmkLJHC8AoIAaBpWnTHAsAwIgKCoCwcL0GQFAQGgrl0F0oiMsQwBcBC1k0LQgLEJZEDBJw6nBH4FQAJ6xbw8XMMQiWqsE2jFJpHBaKQ3lsIIqoMLQyX5bwWDBK4wCOGItCWdJpBYCwhjAOIlUtXg6YlIyYEpaQ/zFK4gKDWCDQmbQeDBMQSXOFgg0EARLCDf1xDBHEmBSpgbVGNNRi2XwBjAAoCJ4Jgc6qoeUnqfwggiGI7BSDIgiKCo6hdbodQGIdpjGOY03BJZkDTKg0RNE1r6qgoMnrQRWAg7i9SNIk9gME4Lg1F4GPtLk%2BRpHECQCP0tQxETTR4504SDA0uWlMMpN6EUJQCC0lRU%2BMNM9K0TODMMnME1w0yKXMCwSKJ4nGV1Zk7NcCK6jCyDID8awnCsJwaDsEC4IQJCqWswu8HlWiTNMCCYEwWDhMjum0QZHBGaQq0RAAHCcXAaJI1ERNR3JrGswau1ywakFJBVmRZVk2ZVdmORASCzAQ0SjeQ9j4EQCN6PdwiiOIL05%2B9agmd9pBznN0QpZLHASWHJlmaqo0p9CqBUHLCtKyrwXq5r2vOD57kG0bMem%2BblvW5Q1dO6tKxcCcrvBtRVGSK7DqOsFGiu3XMucFH1km9ppC6fpYkcGs0sR7vI8iaQ9lOWgWC65nafUL5zBsINOePfn0iF0oxdfT0vodqIB/rmBZk0dGmMUhkx8KMfGExyYZESHzJBxMGCC0QRAhmvMsYDBRvTNmAt4HUxAFyHmfQ8FkzDBzEhXMyEiyUuLZY6xNjbBUvsQ4BBjgEDOBcLciplTPHeF8H4RIIrAj3OOGE1ZawojRBiLEzZMAbhLP8CRkZpFUhpHSPEm4bjMj5L5SM3JeSsjAgQQUwothQQlMuOiboHiYCeCpdUmptR6gNEaE0ZoLRWhtHaB0ToXSOKVM4lUAQ5CijkcYZQchvCWhRMobA%2BoTTeGwJRGicoyzXGZJ6b0vpbgDkDMOMMNgVZSOjB%2BLi05vxsT/IuexfFewVirPCJE8iGxKPMbiNsHYRHdgYv2eiQ5dxkiqcxWpPFbywW0WuVRAjRkjnGQeI8GAiIkQYFeDisFJkznnI098ey6m/mzLxfRQF%2BIhlAuBaxYo7EdggG4hCnjkI%2BLQv4zCdpcKROiR02J8TEmxJSShdJrZcL4UImeC8WyyLnJGcBEMxzWKnJ2TKY50yAIXP4hEQSRThLV1ruHGSckFJMI4esFYxtbI6T0pIE4wZGVMuZcy8hp8nbEtMlfSy%2B8jp3wTs5AeYQX7eTcn5JgAUgohT4HQCKUUYpdTSklQaSqMpZRytYEyRVGAEFKuVEy1Var1VoI1QarV2qdQKvgXqNh%2BpNQKsNZAo1FjqQmqfAqQM5oZQWosAqy08CrWautTaSgdp7Q6n4UAsdjpMFOudS611GCf1kN/Z6v9ZBF0%2BgVXQKxgF/TMPoGaSMwYQ0SFDGGcMwgIwGqDAhrNPAQAcKguBORSF1HSOg1BHbKZ0KFnWpo7MqhUOZnTetg7MHcxoUOmBzNiGtvocLGYYt2ArEJRfElHB26K2AMrVWPctY6wzvrRUq7qWx1pSfQy66uXmQsNHA%2BZsHZUu3pfW9D7pjrXiHYSQQA%3D%3D

Hurricos commented:
> Please define magic number (4250) and then reduce instructions in ISR. Existing: latestVehicleMPH = ONE_HOUR_IN_MICROSECONDS / periodBetweenTicks_us / 4250;
> 
> proposed: latestVehicleMPH = (ONE_HOUR_IN_MICROSECONDS / PULSES_PER_MILE ) / periodBetweenTicks_us;

Done in commit ca3ae67c0649a2d801044a7f7878a54e5af6b55d.
